Improve dev tools in the Nix environment

monacoremo opened the pull request:
@wolfgangwalther has started using the nix tooling and had a plethora of excellent ideas to improve the developer experience. Should not be too difficult to realize, so tracking here:

* [x] add a script to watch for changes (e.g. to re-run tests)
  * [x] add docs
* [x] add a script to run all checks
* [x] add a script to push all artifacts to cachix
* [x] on failure in postgrest-test-spec-all, print the failing postgres version before exiting
* [x] run tests against postgest in ascending/chronological order in postgrest-test-spec-all (was alphabetical)
* [x] add docs on postgrest-style-check

Tbd/potentially for later PRs:
* integrate coverage
* option to preserve temporary dbs and explore them
* script to run postgrest against a temporary db

Edit: tracking those in #1671 

* [x] Refactoring: will also use this opportunity to factor out `set -euo pipefail` from writeShellScriptBin uses and lint bash scripts that we define in Nix.

wolfgangwalther commented:
>     * run tests against postgest in ascending/chronological order in postgrest-test-spec-all (was alphabetical)

Actually my suggestion was to do it in descending order, if I remember correctly. This assumes that one develops with the latest version by default and then goes through `postgrest-test-spec-all`. So you already know it passes with pg13. If `all` goes through backwards, you immediately know the first version it breaks on. If you go upwards you know that it doesn't work in pg 9.5, but don't know where the cut is.

wolfgangwalther commented:
Current situation:
I am in nix-shell. I want to just build the executable without testing it. Ok, I ran `postgrest-test-spec` and have my build, but it took longer than it could. Maybe add a `postgrest-build`? Or am I missing something trivial here?

And now I want to execute the binary to test some cli options. Running `postgrest` tells me it can't be found. Would be good if the built executable would also be on the PATH in nix-shell, right?

wolfgangwalther commented:
It's probably just me missing any stack or cabal background...

So if I enter `cabal run postgrest`, I get what I want. Instead of changing the path, maybe a tool `postgrest` could be just that - an alias on top of `cabal run postgrest`, so that it builds automatically as well?

steve-chavez commented:
> Maybe add a postgrest-build
> maybe a tool postgrest-run

Those would be nice for me as well. I usually do:

- cabal v2-build
- cabal v2-run postgrest support/configs/config

monacoremo commented:
Good idea, especially as cabal is in this weird transition phase with the `v2-*` commands that need to be used.

So let's alias
* `postgrest-build` to `cabal v2-build`
* `postgrest-run` to something like `tests/with_tmp_db cabal v2-run postgrest` (+ one version per Postgres version) - the temporary db can be used if needed, or not
* potentially `clean` and others, tbd

To note: Putting a postgrest build on the nix-shell path immediately would require that we build postgrest through Nix as nix-shell starts up, causing potentially long load times, and would require that nix-shell is restarted when code has changed to get an updated version. Putting the cabal build directory on the path might be tricky, will look into that.

To build a production-ready binary, you would run `nix-build -A postgrestStatic`, but this would be a completely fresh, non-incremental build.

wolfgangwalther commented:
> To note: Putting a postgrest build on the nix-shell path immediately would require that we build postgrest through Nix as nix-shell starts up, causing potentially long load times, and would require that nix-shell is restarted when code has changed to get an updated version. Putting the cabal build directory on the path might be tricky, will look into that.

I don't think we need that, when we have `postgrest-run`. This is much better.

>     * `postgrest-run` to something like `tests/with_tmp_db cabal v2-run postgrest` (+ one version per Postgres version) - the temporary db can be used if needed, or not

This would run the executable with a default config to use the temp db, right?

wolfgangwalther commented on nix/devtools.nix:
```diff
@@ -1,49 +1,96 @@
+    checkedShellScript "postgrest-check"
+      ''
+        ${style}
+        ${tests}/bin/postgrest-test-spec-all
```
Can we flip those two around, and also add `${lint}`?

I think 1. test-spec, 2. lint, 3. style would be the best order. I first want to make sure that the specs pass, then make the linter happy once the code does what it should do, and then let style do some whitespace stuff.

But if the linter fails before the specs are run, doing a `postgrest-watch postgrest-check` might be annoying.

monacoremo replied:
Yes, good idea! I'll also have to add a note that io and memory tests need to be run separately, as they depend on the postgrest executable as built by Nix (at least for now)

wolfgangwalther replied:
Yeah, I just looked into that as well. Can we not just add a `exec ${cabal-install}/bin/cabal v2-exec` before the memory and io tests scripts? This should already fix this, I think. I remember doing that back in the stack days, with stack exec.

monacoremo replied:
Yes, that should work! The profiled build for memory is a bit tricky though, so let's do that in a separate PR.

wolfgangwalther commented:
Ah it's really nice to see shellcheck in immediate action with lots of suggestions. Nice addition! :)

monacoremo commented:
> Ah it's really nice to see shellcheck in immediate action with lots of suggestions. Nice addition! :)

Oh yes, feels good - also to know that all future scripts and changes will be checked and solid!

wolfgangwalther commented:
I see no "shellcheck" tool in devtools.nix - so I assume the check is done at runtime of those tools, right? I follow that about half of our commands are checked in CI as well, but not all of them (release etc.).

monacoremo commented:
> I see no "shellcheck" tool in devtools.nix - so I assume the check is done at runtime of those tools, right? I follow that about half of our commands are checked in CI as well, but not all of them (release etc.).

I added 'checkedShellScript' in an overlay, which is now used everywhere in our nix setup. It checks the scripts at nix evaluation time, so only at first use e.g. when you launch nix shell.

monacoremo commented:
I will need to split out the styling/linting scripts, as at this point we are pulling in too many dependencies for our poor styling job...

wolfgangwalther commented:
Ah, I see. So `nix-shell` evaluates everything, right? Is the same true for those tasks that are triggered in CI? Or could we maybe just add another CI step that just launches the `nix-shell` and exit to trigger those?

monacoremo commented:
> Ah, I see. So `nix-shell` evaluates everything, right? Is the same true for those tasks that are triggered in CI? Or could we maybe just add another CI step that just launches the `nix-shell` and exit to trigger those?

The exact same checks happen when we do `nix-env -iA ...` or `nix-build` in CI - so it's all already covered!

The checks are not re-run when the exact scriot is already im the nix store or we pull from cachix, but everything in there is guaranteed to have been checked.

wolfgangwalther commented on nix/README.md:
```diff
@@ -161,6 +163,20 @@ $ nix-shell --run postgrest-style
+Tools like `postgrest-build`, `postgrest-run` etc. are simple wrappers around
+`cabal` and should do what you expect. `postgrest-check` runs most checks that will
+also run in CI, with the exception of the IO and Memory checks that need to be run
+separarely.
```
```suggestion
separately.
```

wolfgangwalther commented on nix/README.md:
```diff
@@ -161,6 +163,20 @@ $ nix-shell --run postgrest-style
+
+`postgrest-watch` takes a command as an argument that it will rerun if any source
+file is changed. For example, `postgrest-watch postgrest-test-spec-all` will re-run
+the full spec test suite againsg all PostgreSQL version on every change.
```
```suggestion
the full spec test suite against all PostgreSQL versions on every change.
```

monacoremo replied:
Thanks! Completed this on my phone, works surprisingly well! But it's not so great for proof-reading :)

wolfgangwalther commented on nix/README.md:
```diff
@@ -161,6 +163,20 @@ $ nix-shell --run postgrest-style
+also run in CI, with the exception of the IO and Memory checks that need to be run
+separarely.
+
+`postgrest-watch` takes a command as an argument that it will rerun if any source
```
```suggestion
`postgrest-watch` takes a command as an argument that it will re-run if any source
```
(same as below)

wolfgangwalther commented:
I would like to give this a test-run before merging by rebasing one of my current branches on it and using the tools. Can we squash all those commits in here into one? I can also do it, while I'm on it, if you want me to. I should be able to force push on this branch. Learned by accident :D

monacoremo commented:
> Can we squash all those commits in here into one? I can also do it, while I'm on it, if you want me to. I should be able to force push on this branch. Learned by accident :D

Yes, please do! Don't have access to my machine right now. Let's see those superpowers :-)

wolfgangwalther commented:
So here are a couple of notes. Maybe some of those can be fixed easily, others might take more work and be done later.

- I can't seem to exit `postgrest-watch`. When I hit CTRL+C, it restarts the watched program, but does not pass through sigint (or whatever it is). Could be specific to my setup (terminal) as well, could you give it a try?

- `postgrest-test-spec-all` outputs `Failed` after every run - even if successful. I think the `trap exit` needs to check for exit code as well.

```
Done running spec against postgresql-9.5.23.

Failed on postgresql-9.5.23
```

We might not need to check the return code at all, but just echo `Done` with the trap each time. Done does not imply success and the error will be shown above anyway?

- auto-complete does not work when adding a command after `postgrest-watch`. That would be really cool, I cant remember all those commands that we now have :)

- it seems like `postgrest-build` and `postgrest-run` on the one hand, and `postgrest-test-...` on the other have different build settings. That leads to repeated builds, when switching between those commands. When I repeat `postgrest-run`, it just says "Up to date" and starts immediately. But when I run `postgrest-run` and `postgrest-test-spec` after each other, it always rebuilds. Very quickly, but it's never "Up to date".

- the `postgrest-style`, `postgrest-lint` and the stylecheck are not available in nix-shell anymore? Probably a result of the refactor because of dependencies?

monacoremo commented:
Thank you for the thorough testing @wolfgangwalther !

>     * I can't seem to exit `postgrest-watch`. When I hit CTRL+C, it restarts the watched program, but does not pass through sigint (or whatever it is). Could be specific to my setup (terminal) as well, could you give it a try?

Same issue here - solved by not handling new files for now, which requires a restart of postgrest-watch. Will look further into a proper solution

>     * `postgrest-test-spec-all` outputs `Failed` after every run - even if successful. I think the `trap exit` needs to check for exit code as well.

Fixed by resetting the trap

>     * auto-complete does not work when adding a command after `postgrest-watch`. That would be really cool, I cant remember all those commands that we now have :)

That's really unfortunate, just noticed myself. Unfortunately this is beyond my bash-fu right now, will also need to solve later
 
>     * it seems like `postgrest-build` and `postgrest-run` on the one hand, and `postgrest-test-...` on the other have different build settings. That leads to repeated builds, when switching between those commands. When I repeat `postgrest-run`, it just says "Up to date" and starts immediately. But when I run `postgrest-run` and `postgrest-test-spec` after each other, it always rebuilds. Very quickly, but it's never "Up to date".

True, I missed the FailOnWarn flag. Should be fixed now I believe

>     * the `postgrest-style`, `postgrest-lint` and the stylecheck are not available in nix-shell anymore? Probably a result of the refactor because of dependencies?

Fixed

wolfgangwalther commented:
> True, I missed the FailOnWarn flag. Should be fixed now I believe

Hm, no I'm still getting ` - postgrest-7.0.1 (lib) (configuration changed)`, whenever I change between the two commands.



> Same issue here - solved by not handling new files for now, which requires a restart of postgrest-watch. Will look further into a proper solution

I noticed something else as well: When I do a watch of something that runs a bit longer (`test-spec-all` maybe) and I do a change while it is running, the tests are not restarted and also not run again after finishing. That's confusing, because it force me to restart/trigger the tests manually sometimes. Could be that I save multiple files and the tests already start at the first file, or that shortly after saving I see a typo and fix it.

wolfgangwalther commented:
> > True, I missed the FailOnWarn flag. Should be fixed now I believe
> 
> Hm, no I'm still getting ` - postgrest-7.0.1 (lib) (configuration changed)`, whenever I change between the two commands.

I have just confirmed, that this is because of `--test-show-detail=direct`. It seems to be possible to add this to the `v2-build` and `v2-run` commands without problem, even though it is probably unused. So we can do that easily.

wolfgangwalther commented:
Looks good! ready to merge from your side?

monacoremo commented:
Just confirmed that both issues from your last comments are fixed - good to go!

Thank you for reviewing and debugging cabal :-)

wolfgangwalther commented:
Awesome job remo, I'm looking forward to working with that!
